packages/middleware/src/loggerManager.ts

```diff
@@ -0,0 +1,108 @@
+import type { Response } from "express";
+import { LoggerInterface } from "@vue-storefront/logger";
+import type {
+  AlokaiContainer,
+  IntegrationContext,
+  MiddlewareConfig,
+} from "./types";
+
+type LoggerInstance = LoggerInterface;
+const GLOBAL_MIDDLEWARE_CFG_KEY = Symbol("GLOBAL_MIDDLEWARE_CFG_KEY");
+export class LoggerManager<
+  TLoggerConfig = any,
+  TIntegrationContext extends Record<string, IntegrationContext> = any
+> {
+  private instances: Record<string | symbol, LoggerInstance> = {};
+
+  /** Creating minimal amount of needed Logger instances * */
+  constructor(
+    config: MiddlewareConfig<TIntegrationContext>, // Whole middleware.config.js object
+    buildLogger: (logger: TLoggerConfig) => LoggerInstance
+  ) {
+    // Global Logger
+    this.instances[GLOBAL_MIDDLEWARE_CFG_KEY] = buildLogger(
+      config.logger as TLoggerConfig // logger is on same level as integrations
+    );
+    // Logger for every integration having custom logger config
+    for (const [integrationName, integrationEntry] of Object.entries(
+      config.integrations
+    )) {
+      if (this.hasCustomLoggerConfig(integrationEntry)) {
+        // every integration inside `configuration` object can have own logger too
+        this.instances[integrationName] = buildLogger(
+          integrationEntry.logger as TLoggerConfig
+        );
+      }
+    }
+  }
+
+  // eslint-disable-next-line class-methods-use-this
+  private hasCustomLoggerConfig(integrationEntry) {
+    return Boolean(integrationEntry.logger);
+  }
+
+  // getting reference to instance of logger by integration tag name
+  public get(integrationName) {
+    return (
+      this.instances[integrationName] ||
+      this.instances[GLOBAL_MIDDLEWARE_CFG_KEY]
+    );
+  }
+
+  // getting reference to global instance of logger
+  public getGlobal() {
+    return this.instances[GLOBAL_MIDDLEWARE_CFG_KEY];
+  }
+}
+
+export function getLogger(
+  source: AlokaiContainer | Response | { res: Response }
+): LoggerInstance {
+  if (!source) {
+    throw new Error("Parameter for getLogger function not provided");
+  }
+  if ("logger" in source) {
+    return source.logger;
+  }
+  let base: Response;
+  if ("res" in source) {
+    base = source.res;
+  } else {
+    base = source;
+  }
+  return base.locals.alokai.logger;
+}
+
+const methodsToOverwrite = [
+  "emergency",
+  "alert",
+  "critical",
+  "error",
+  "warning",
+  "notice",
+  "info",
+  "debug",
+];
+
+export function wrapLogger(
+  logger: LoggerInstance,
+  metadataGetter: (metadata: Record<string, any>) => Record<string, any>
```

Comment:
Why do we need a function? Why can't we simply merge metadata? Isn't it an overengineering? The function produces data that we merge anyway, why not inject the data itself?

Reply:
Take a look at the following middleware:
```ts
export function prepareLogger(loggerManager: LoggerManager) {
  return function (req: Request, res: Response, next: NextFunction) {
    if (!res.locals.alokai) {
      res.locals.alokai = {};
    }
    const logger = loggerManager.get(req.params.integrationName);
    const proxiedLogger = injectMetadata(logger, (metadata) => {
      return {
        ...res.locals.alokai.metadata,
        ...metadata,
        scope: {
          ...res.locals.alokai.metadata?.scope,
          ...metadata?.scope,
        },
      };
    });
    res.locals.alokai.logger = proxiedLogger;

    next();
  };
}
```

This middleware is the second one that's called when someone calls any of our endpoints (first is creating res.locals.alokai.metadata object). In that moment, `res.locals.alokai.metadata` doesn't have many information. We adjust it's value in different moments of request's flow. If we passed ` ...res.locals.alokai.metadata` here, without getter, then it would be instantly resolved (won't have data added later in request's flow). Getter assures us, we always get actual value of `res.locals.alokai.metadata` (including fields added later like functionName, integrationName, extensionName).